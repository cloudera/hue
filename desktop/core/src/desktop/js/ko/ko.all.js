// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as ko from 'knockout';
import 'ko/ko.init';
import komapping from 'knockout.mapping';
import 'knockout-switch-case';
import 'knockout-sortable';
import 'knockout.validation';

import 'ext/ko.editable.custom';
import 'ext/ko.selectize.custom';

import 'ko/bindings/charts/mapchart/ko.mapChart';
import 'ko/bindings/charts/ko.barChart';
import 'ko/bindings/charts/ko.leafletMapChart';
import 'ko/bindings/charts/ko.lineChart';
import 'ko/bindings/charts/ko.partitionChart';
import 'ko/bindings/charts/ko.pieChart';
import 'ko/bindings/charts/ko.scatterChart';
import 'ko/bindings/charts/ko.timelineChart';

import 'ko/bindings/ko.vue.js';

import 'ko/bindings/ace/ko.aceEditor';
import 'ko/bindings/ace/ko.aceResizer';
import 'ko/bindings/ko.appAwareTemplateContextMenu';
import 'ko/bindings/ko.assistFileDraggable';
import 'ko/bindings/ko.assistFileDroppable';
import 'ko/bindings/ko.assistVerticalResizer';
import 'ko/bindings/ko.attachViewModelToElementData';
import 'ko/bindings/ko.augmentHtml';
import 'ko/bindings/ko.autocomplete';
import 'ko/bindings/ko.autogrowInput';
import 'ko/bindings/ko.blurHide';
import 'ko/bindings/ko.bytesize';
import 'ko/bindings/ko.chosen';
import 'ko/bindings/ko.clearable';
import 'ko/bindings/ko.clickForAceFocus';
import 'ko/bindings/ko.clickToCopy';
import 'ko/bindings/ko.clipboard';
import 'ko/bindings/ko.codemirror';
import 'ko/bindings/ko.contextMenu';
import 'ko/bindings/ko.contextSubMenu';
import 'ko/bindings/ko.datalist';
import 'ko/bindings/ko.datepicker';
import 'ko/bindings/ko.dateRangePicker';
import 'ko/bindings/ko.dblClick';
import 'ko/bindings/ko.delayedOverflow';
import 'ko/bindings/ko.dockable';
import 'ko/bindings/ko.documentChooser';
import 'ko/bindings/ko.documentContextPopover';
import 'ko/bindings/ko.draggableText';
import 'ko/bindings/ko.dropdown';
import 'ko/bindings/ko.dropzone';
import 'ko/bindings/ko.duration';
import 'ko/bindings/ko.ellipsis';
import 'ko/bindings/ko.fadeVisible';
import 'ko/bindings/ko.fetchMore';
import 'ko/bindings/ko.fileChooser';
import 'ko/bindings/ko.foreachVisible';
import 'ko/bindings/ko.fresherEditor';
import 'ko/bindings/ko.hdfsAutocomplete';
import 'ko/bindings/ko.hdfsTree';
import 'ko/bindings/ko.highlight';
import 'ko/bindings/ko.hiveChooser';
import 'ko/bindings/ko.html';
import 'ko/bindings/ko.hueach';
import 'ko/bindings/ko.hueCheckAll';
import 'ko/bindings/ko.hueCheckbox';
import 'ko/bindings/ko.hueChecked';
import 'ko/bindings/ko.hueLink';
import 'ko/bindings/ko.hueSpinner';
import 'ko/bindings/ko.impalaDagre';
import 'ko/bindings/ko.jHueRowSelector';
import 'ko/bindings/ko.logResizer';
import 'ko/bindings/ko.logScroller';
import 'ko/bindings/ko.medium';
import 'ko/bindings/ko.moment';
import 'ko/bindings/ko.momentFromNow';
import 'ko/bindings/ko.multiCheck';
import 'ko/bindings/ko.multiCheckForeachVisible';
import 'ko/bindings/ko.multiClick';
import 'ko/bindings/ko.multiLineEllipsis';
import 'ko/bindings/ko.numberFormat';
import 'ko/bindings/ko.numericTextInput';
import 'ko/bindings/ko.onClickOutside';
import 'ko/bindings/ko.oneClickSelect';
import 'ko/bindings/ko.parseArguments';
import 'ko/bindings/ko.publish';
import 'ko/bindings/ko.reactWrapper';
import 'ko/bindings/ko.readOnlyAce';
import 'ko/bindings/ko.resizable';
import 'ko/bindings/ko.select2';
import 'ko/bindings/ko.simplesize';
import 'ko/bindings/ko.slider';
import 'ko/bindings/ko.slideVisible';
import 'ko/bindings/ko.solrChooser';
import 'ko/bindings/ko.spinEdit';
import 'ko/bindings/ko.splitDraggable';
import 'ko/bindings/ko.splitFlexDraggable';
import 'ko/bindings/ko.sqlContextPopover';
import 'ko/bindings/ko.stickVertical';
import 'ko/bindings/ko.storageContextPopover';
import 'ko/bindings/ko.stretchDown';
import 'ko/bindings/ko.tagEditor';
import 'ko/bindings/ko.tagsNotAllowed';
import 'ko/bindings/ko.templateContextMenu';
import 'ko/bindings/ko.templatePopover';
import 'ko/bindings/ko.textSqueezer';
import 'ko/bindings/ko.toggle';
import 'ko/bindings/ko.toggleOverflow';
import 'ko/bindings/ko.tooltip';
import 'ko/bindings/ko.truncatedText';
import 'ko/bindings/ko.typeahead';
import 'ko/bindings/ko.verticalSlide';
import 'ko/bindings/ko.visibleOnHover';

import 'ko/components/appSwitcher/ko.appSwitcher';
import 'ko/components/assist/ko.assistDashboardPanel';
import 'ko/components/assist/ko.assistDbPanel';
import 'ko/components/assist/ko.assistDocumentsPanel';
import 'ko/components/assist/ko.assistEditorContextPanel';
import 'ko/components/assist/ko.assistFunctionsPanel';
import 'ko/components/assist/ko.assistGitPanel';
import 'ko/components/assist/ko.assistHBasePanel';
import 'ko/components/assist/ko.assistLangRefPanel';
import 'ko/components/assist/ko.assistPanel';
import 'ko/components/assist/ko.assistSchedulePanel';
import 'ko/components/assist/ko.assistStoragePanel';
import 'ko/components/assist/ko.rightAssistPanel';
import 'ko/components/contextPopover/ko.contextPopover';
import 'ko/components/contextPopover/ko.quickQueryContext';
import 'ko/components/contextPopover/ko.popoverOnHover';
import 'ko/components/doc/ko.shareDocModal';
import 'ko/components/simpleAceEditor/ko.simpleAceEditor';

import 'ko/components/ko.catalogEntriesList';
import 'ko/components/ko.contextSelector';
import 'ko/components/ko.createDirectoryModal';
import 'ko/components/ko.deleteDocModal';
import 'ko/components/ko.dropDown';
import 'ko/components/ko.editorDroppableMenu';
import 'ko/components/ko.executionAnalysis';
import 'ko/components/ko.favoriteApp';
import 'ko/components/ko.fieldSamples';
import 'ko/components/ko.globalSearch';
import 'ko/components/ko.historyPanel';
import 'ko/components/ko.importDocumentsModal';
import 'ko/components/ko.inlineAutocomplete';
import 'ko/components/ko.jobBrowserLinks';
import 'ko/components/ko.navProperties';
import 'ko/components/ko.navTags';
import 'ko/components/ko.pollingCatalogEntriesList';
import 'ko/components/ko.quickQueryAction';
import 'ko/components/ko.sentryPrivileges';
import 'ko/components/ko.sessionAuthModal';
import 'ko/components/ko.sessionPanel';
import 'ko/components/ko.shareGistModal';
import 'ko/components/ko.sqlColumnsTable';

// TODO: Move to about app when it has it's own webpack entry
import 'apps/about/components/ko.hueConfigTree';
import 'apps/about/components/ko.connectorsConfig';

import 'ko/extenders/ko.maxLength';
import 'ko/extenders/ko.numeric';
import 'ko/extenders/ko.toJson';

import 'ko/observables/ko.observableArrayDefault';
import 'ko/observables/ko.observableDefault';

window.ko = ko;
window.ko.mapping = komapping;

export default ko;
