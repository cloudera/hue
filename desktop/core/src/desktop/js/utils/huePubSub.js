// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Based on original pub/sub implementation from http://davidwalsh.name/pubsub-javascript

const topics = {};
const hOP = topics.hasOwnProperty;

const huePubSub = {
  subscribe: function (topic, listener, app) {
    if (!hOP.call(topics, topic)) {
      topics[topic] = [];
    }

    const index =
      topics[topic].push({
        listener: listener,
        app: app,
        status: 'running'
      }) - 1;

    return {
      remove: function () {
        delete topics[topic][index];
      }
    };
  },
  removeAll: function (topic) {
    topics[topic] = [];
  },
  subscribeOnce: function (topic, listener, app) {
    const ephemeral = this.subscribe(
      topic,
      function () {
        listener.apply(listener, arguments);
        ephemeral.remove();
      },
      app
    );
  },
  publish: function (topic, info) {
    if (!hOP.call(topics, topic)) {
      return;
    }

    topics[topic].forEach(item => {
      if (item.status === 'running') {
        item.listener(info);
      }
    });
  },
  getTopics: function () {
    return topics;
  },
  pauseAppSubscribers: function (app) {
    if (app) {
      Object.keys(topics).forEach(topicName => {
        topics[topicName].forEach(topic => {
          if (
            typeof topic.app !== 'undefined' &&
            topic.app !== null &&
            (topic.app === app || topic.app.split('-')[0] === app)
          ) {
            topic.status = 'paused';
          }
        });
      });
    }
  },
  resumeAppSubscribers: function (app) {
    if (app) {
      Object.keys(topics).forEach(topicName => {
        topics[topicName].forEach(topic => {
          if (
            typeof topic.app !== 'undefined' &&
            topic.app !== null &&
            (topic.app === app || topic.app.split('-')[0] === app)
          ) {
            topic.status = 'running';
          }
        });
      });
    }
  },
  clearAppSubscribers: function (app) {
    if (app) {
      Object.keys(topics).forEach(topicName => {
        topics[topicName] = topics[topicName].filter(obj => {
          return obj.app !== app;
        });
      });
    }
  }
};

export default huePubSub;
