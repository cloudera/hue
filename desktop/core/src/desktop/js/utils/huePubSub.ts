// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Based on original pub/sub implementation from http://davidwalsh.name/pubsub-javascript

/* eslint-disable @typescript-eslint/no-explicit-any */

interface Listener<T> {
  (data: T): void;
}

enum PubSubState {
  RUNNING,
  PAUSED
}

const topics: {
  [topic: string]: {
    listener: Listener<any>;
    app?: string;
    state: PubSubState;
  }[];
} = {};

const hOP = topics.hasOwnProperty;

export interface HueSubscription {
  remove: () => void;
}

const subscribe = <T = any>(
  topic: string,
  listener: Listener<T>,
  app?: string
): HueSubscription => {
  if (!hOP.call(topics, topic)) {
    topics[topic] = [];
  }

  const index =
    topics[topic].push({
      listener: listener,
      app: app,
      state: PubSubState.RUNNING
    }) - 1;

  return {
    remove: () => {
      delete topics[topic][index];
    }
  };
};

const removeAll = (topic: string): void => {
  topics[topic] = [];
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
const publish = <T = any>(topic: string, info?: T): void => {
  if (!hOP.call(topics, topic)) {
    return;
  }

  topics[topic].forEach(item => {
    if (item.state === PubSubState.RUNNING) {
      item.listener(info);
    }
  });
};

const subscribeOnce = <T = any>(
  topic: string,
  listener: Listener<T>,
  app: string
): HueSubscription => {
  const ephemeral = subscribe(
    topic,
    arg => {
      listener(arg);
      ephemeral.remove();
    },
    app
  );
  return ephemeral;
};

const getTopics = (): typeof topics => topics;

const setStatusForApp = (state: PubSubState, app?: string): void => {
  if (app) {
    Object.keys(topics).forEach(topicName => {
      topics[topicName].forEach(topic => {
        if (topic.app && (topic.app === app || topic.app.split('-')[0] === app)) {
          topic.state = state;
        }
      });
    });
  }
};

const pauseAppSubscribers = (app?: string): void => {
  setStatusForApp(PubSubState.PAUSED, app);
};

const resumeAppSubscribers = (app?: string): void => {
  setStatusForApp(PubSubState.RUNNING, app);
};

const clearAppSubscribers = (app?: string): void => {
  if (app) {
    Object.keys(topics).forEach(topicName => {
      topics[topicName] = topics[topicName].filter(obj => {
        return obj.app !== app;
      });
    });
  }
};

export default {
  clearAppSubscribers,
  getTopics,
  pauseAppSubscribers,
  publish,
  removeAll,
  resumeAppSubscribers,
  subscribe,
  subscribeOnce
};
