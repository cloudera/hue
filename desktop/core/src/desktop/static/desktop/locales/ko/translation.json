{
  "Add": "추가",
  "aggregate": "집계",
  "Aggregate": "집계",
  "alias": "별칭",
  "All": "모두",
  "and": "및",
  "Apps": "앱",
  "Assist": "지원",
  "Assistant": "도우미",
  "Back": "뒤로",
  "Browsers": "브라우저",
  "Bundle": "Bundle",
  "Cancel": "취소",
  "Change": "변경",
  "Clipboard": "클립보드",
  "Close": "닫기",
  "Clusters": "클러스터",
  "Columns": "열",
  "condition": "조건",
  "Create": "생성",
  "Created": "생성됨",
  "CSV": "CSV",
  "cte": "cte",
  "CTEs": "CTE",
  "Dashboard": "대시보드",
  "database": "데이터베이스",
  "Databases": "데이터베이스",
  "Date": "날짜",
  "default": "기본값:",
  "Delete": "삭제",
  "Description": "설명",
  "Details": "세부 정보",
  "Directory": "디렉토리",
  "Documentation": "문서",
  "Documents": "문서",
  "Editor": "편집기",
  "Excel": "Excel",
  "Export": "내보내기",
  "Failed": "실패함",
  "Field": "필드",
  "File": "파일",
  "Files": "파일",
  "filter": "필터",
  "Filter": "필터",
  "Format": "형식",
  "Functions": "함수",
  "Group": "그룹",
  "HBase": "HBase",
  "Help": "도움말",
  "Hive": "Hive",
  "Identifiers": "식별자",
  "Impala": "Impala",
  "Import": "가져오기",
  "Indexes": "인덱스",
  "Insert": "커서 위치에 ",
  "Jobs": "Job",
  "join": "가입",
  "Key": "키",
  "keyword": "키워드",
  "Keywords": "키워드",
  "Metrics": "메트릭",
  "Modify": "수정",
  "More": "더",
  "Name": "이름",
  "Next": "다음",
  "No": "아니요",
  "Notebook": "노트북",
  "of": "/",
  "Open": "열기",
  "Options": "옵션",
  "Order": "순서",
  "other": "기타",
  "Output": "출력",
  "Owner": "소유자",
  "Partitions": "파티션",
  "Password": "암호",
  "Permissions": "권한",
  "Popular": "인기",
  "Popularity": "인기도",
  "Project": "프로젝트",
  "Queries": "쿼리",
  "Read": "읽기",
  "Refresh": "새로 고침",
  "Remove": "제거",
  "Report": "보고서",
  "Running": "실행 중",
  "Sample": "샘플",
  "sample": "샘플",
  "Samples": "샘플",
  "Schedule": "예약",
  "Search": "검색",    
  "Settings": "설정",
  "Share": "공유",
  "Sharing": "공유",
  "Size": "크기",
  "Sources": "소스",
  "SQL": "SQL",
  "Start": "시작",
  "Statement": "명령문",
  "Stats": "상태",
  "Stopped": "중지됨",
  "table": "테이블",
  "Table": "테이블",
  "Tables": "테이블",
  "Terms": "용어",
  "Type": "유형",
  "UDFs": "UDF",
  "Undo": "실행 취소",
  "Username": "사용자 이름",
  "Value": "값",
  "Values": "값",
  "variable": "변수",
  "Variables": "변수",
  "View": "보기",
  "view": "보기",
  "Views": "보기",
  "virtual": "가상",
  "Workflow": "Workflow",
  "Yes": "예"
}