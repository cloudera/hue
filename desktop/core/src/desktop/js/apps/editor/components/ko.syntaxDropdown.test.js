// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { SHOW_EVENT, SHOWN_EVENT, SYNTAX_DROPDOWN_ID } from './ko.syntaxDropdown';
import huePubSub from 'utils/huePubSub';

describe('ko.syntaxDropdown.js', () => {
  const publishShowEvent = () => {
    huePubSub.publish(SHOW_EVENT, {
      data: {
        text: 'floo',
        expected: [{ text: 'foo' }],
        ruleId: 123
      },
      source: {
        left: 10,
        bottom: 20
      }
    });
  };

  it('should render component on show event', async () => {
    expect(document.querySelectorAll(`#${SYNTAX_DROPDOWN_ID}`)).toHaveLength(0);

    publishShowEvent();

    expect(document.querySelectorAll(`#${SYNTAX_DROPDOWN_ID}`)).toHaveLength(1);
    expect(window.document.documentElement.outerHTML).toMatchSnapshot();
  });

  it('should match snapshot', async () => {
    const shownPromise = new Promise(resolve => {
      huePubSub.subscribeOnce(SHOWN_EVENT, resolve);
    });
    publishShowEvent();

    await shownPromise;

    expect(window.document.documentElement.outerHTML).toMatchSnapshot();
  });
});
