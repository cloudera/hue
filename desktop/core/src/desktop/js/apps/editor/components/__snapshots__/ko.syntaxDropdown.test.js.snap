// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ko.syntaxDropdown.js should match snapshot 1`] = `
"<html><head></head><body><div id=\\"sqlSyntaxDropdown\\" data-bind=\\"component: { name: 'sql-syntax-dropdown', params: $data }\\">
<div class=\\"hue-syntax-drop-down\\" data-bind=\\"
  style: {
    'left': left() + 'px',
    'top': top() + 'px'
  },
  component: {
    name: 'hue-drop-down',
    params: {
      value: selected,
      entries: expected,
      foreachVisible: false,
      searchable: false,
      showOnInit: true,
      menuOnly: true
    }
  }
\\" style=\\"left: 10px; top: 20px;\\">
  <!-- ko if: !menuOnly && (!dropDownVisible() || !searchable) --><!-- /ko -->
  <!-- ko if: !menuOnly && (dropDownVisible() && searchable) --><!-- /ko -->
  <div class=\\"hue-drop-down-container open\\" data-bind=\\"css: { 'open' : dropDownVisible, 'hue-drop-down-fixed': fixedPosition, 'hue-drop-down-container-searchable': searchable }, dropDownKeyUp: { onEsc: onEsc, onEnter: onEnter, dropDownVisible: dropDownVisible }\\">
    <div style=\\"overflow-y: auto;\\" class=\\"dropdown-menu\\" data-bind=\\"visible: filteredEntries().length > 0\\">
      <!-- ko if: foreachVisible --><!-- /ko -->
      <!-- ko ifnot: foreachVisible -->
      <ul class=\\"hue-inner-drop-down\\" data-bind=\\"foreach: filteredEntries\\">
        <!-- ko if: typeof $data.divider !== 'undefined' && $data.divider --><!-- /ko -->
        <!-- ko if: typeof $data.divider === 'undefined' || !$data.divider -->
        <li><a href=\\"javascript:void(0)\\" data-bind=\\"text: $data &amp;&amp; typeof $data[$parent.labelAttribute] !== 'undefined' ? $data[$parent.labelAttribute] : $data, click: function () { let previous = $parent.value(); $parent.value($data); $parent.onSelect($data, previous); }\\">foo</a></li>
        <!-- /ko -->
      
        <!-- ko if: typeof $data.divider !== 'undefined' && $data.divider -->
        <li class=\\"divider\\"></li>
        <!-- /ko -->
        <!-- ko if: typeof $data.divider === 'undefined' || !$data.divider --><!-- /ko -->
      
        <!-- ko if: typeof $data.divider !== 'undefined' && $data.divider --><!-- /ko -->
        <!-- ko if: typeof $data.divider === 'undefined' || !$data.divider -->
        <li><a href=\\"javascript:void(0)\\" data-bind=\\"text: $data &amp;&amp; typeof $data[$parent.labelAttribute] !== 'undefined' ? $data[$parent.labelAttribute] : $data, click: function () { let previous = $parent.value(); $parent.value($data); $parent.onSelect($data, previous); }\\">Ignore this type of error</a></li>
        <!-- /ko -->
      </ul>
      <!-- /ko -->
    </div>
  </div>
</div>
</div></body></html>"
`;

exports[`ko.syntaxDropdown.js should render component on show event 1`] = `"<html><head></head><body><div id=\\"sqlSyntaxDropdown\\" data-bind=\\"component: { name: 'sql-syntax-dropdown', params: $data }\\"></div></body></html>"`;
